### PR TITLE
Use Bank module tokens as deposit tokens instead of CW20

### PR DESCRIPTION
**Pull request by Overseven**

### CODE REVIEW

File: cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs [hunk 1]
```diff
@@ -115,48 +122,33 @@ pub fn try_handle_unstake_response(
...
 ) -> Result<(), ContractError> {
-    // assert message funds
-    if token_amount.is_zero() && unstake_response.reinit_unstake_id != 0 {
+    if info.funds.len() == 0 && unstake_response.reinit_unstake_id != 0 {
         return Err(ContractError::CustomError(
             "Unstake response: reinit_unstake_id != 0, but message have no tokens".to_string(),
         ));
     }
-    if !token_amount.is_zero() {
+    if info.funds.len() == 1 {
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs]: this `if` can be removed by placing the `if info.funds.len() > 1` check before it

File: cosmwasm/contracts/yield-ward/src/execute/stake.rs [hunk 2]
```diff
@@ -1,23 +1,24 @@
...
-    // }
-    // let coin = info.funds.first().unwrap();
+    if info.funds.len() != 1 {
+        return Err(ContractError::CustomError(
+            "Init stake message must have one type of coins as funds".to_string(),
+        ));
+    }
+    let Coin {
+        amount: token_amount,
+        denom: token_denom,
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/stake.rs:19]: the reference sign in `&token_denom` can be removed (3 places in this file)

File: cosmwasm/contracts/yield-ward/src/helpers.rs [hunk 3]
```diff
@@ -61,19 +64,12 @@ pub fn find_token_by_message_source(
...
-        .iter()
-        .find(|(_, config)| &config.lpt_denom == lp_token_denom)
-        .cloned()
-        .ok_or(ContractError::UnknownLpToken(lp_token_denom.clone()))
+    lpt_address: &Addr,
+) -> Result<TokenDenom, ContractError> {
+    Ok(TOKEN_DENOM_BY_LPT_ADDRESS
+        .load(deps.storage, lpt_address)
+        .ok()
+        .ok_or(ContractError::UnknownLpToken(lpt_address.to_string()))?)
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/helpers.rs]:
```
TOKEN_DENOM_BY_LPT_ADDRESS
        .load(deps.storage, lpt_address)
        .map_err(|_| ContractError::UnknownLpToken(lpt_address.to_string()))
```

File: cosmwasm/contracts/yield-ward/src/query.rs [hunk 4]
```diff
@@ -86,3 +87,18 @@ pub fn query_all_tokens_denoms_by_source(deps: Deps) -> StdResult<GetTokenDenomB
...
+    deps: Deps,
+) -> StdResult<GetTokenDenomByLptAddressResponse> {
+    let tokens_denoms: StdResult<Vec<_>> = TOKEN_DENOM_BY_LPT_ADDRESS
+        .range(deps.storage, None, None, Order::Ascending)
+        .collect();
+
+    let tokens_denoms: Vec<_> = tokens_denoms?
+        .into_iter()
+        .map(|(lpt_address, token_denom)| (lpt_address, token_denom))
+        .collect();
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/query.rs]: I don't think it's a good idea to collect twice (even in view method)

File: cosmwasm/contracts/yield-ward/src/helpers.rs (hunk 3 above)
Comment (optifat) [reply cosmwasm/contracts/yield-ward/src/helpers.rs]: The same in `try_receive_cw20` `let action_msg: Cw20ActionMsg = ...` (can't mark it cause it out of changes scope)

File: cosmwasm/contracts/yield-ward/src/execute/reinit.rs [hunk 5]
```diff
@@ -84,33 +80,41 @@ pub fn handle_reinit(
...
+    info: MessageInfo,
     source_chain: String,
     source_address: String,
     payload: &[u8],
-    token_amount: Uint128,
 ) -> Result<Response, ContractError> {
-    if token_amount.is_zero() {
+    if info.funds.len() != 1 || info.funds.first().unwrap().amount.is_zero() {
         return Err(ContractError::CustomError(
             "Reinit message must have one type of coins as funds".to_string(),
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/reinit.rs:88]: Wouldn't it be better to make an error message of each condition its own?

File: cosmwasm/contracts/yield-ward/src/execute/common.rs [hunk 6]
```diff
@@ -13,7 +11,17 @@ pub fn create_cw20_mint_msg(
...
+    .ok()?;
+
+    Some(WasmMsg::Execute {
+        contract_addr: cw20_address.to_string(),
+        msg,
+        funds: vec![],
+    })
+}
+
+pub fn create_cw20_burn_msg(cw20_address: &Addr, amount: Uint128) -> Option<WasmMsg> {
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/common.rs]: Tbh returning `Option` here is kinda weird: you're doing `ok_or(error)` later anyway. Why not change the return type to `Result` and `to_json_binary(...).map_err(new_error)`?

File: cosmwasm/contracts/yield-ward/src/execute/receive_cw20.rs [hunk 7]
```diff
@@ -20,40 +18,13 @@ pub fn try_receive_cw20(
     let user = deps.api.addr_validate(&msg.sender)?;
 
     match action_msg {
-        Cw20ActionMsg::Stake {
-            deposit_token_denom,
-        } => {
-            assert_cw20_deposit_token_address(deps.as_ref(), &deposit_token_denom, &info.sender)?;
+        Cw20ActionMsg::Unstake {} => {
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/receive_cw20.rs]: redundant `{}`

File: cosmwasm/contracts/yield-ward/src/execute/response_stake.rs [hunk 8]
```diff
@@ -147,47 +135,25 @@ fn ensure_stake_response_is_valid(
             ));
         }
     }
-    if !token_amount.is_zero() {
+    if info.funds.len() == 1 {
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/response_stake.rs]: this `if` can be removed by placing the `if info.funds.len() > 1` check before it

File: cosmwasm/contracts/yield-ward/src/execute/reinit.rs [hunk 9]
```diff
@@ -18,29 +18,26 @@ pub fn try_reinit(
...
     deposit_token_denom: &TokenDenom,
-    deposit_token_address: &Addr,
     token_amount: Uint128,
     reinit_unstake_id: &u64,
     mut stake_stats: StakeStatsItem,
-) -> Result<(WasmMsg, Event), ContractError> {
+) -> Result<(BankMsg, Event), ContractError> {
     let mut unstake_item = UNSTAKES.load(
         deps.storage,
         (&deposit_token_denom, reinit_unstake_id.clone()),
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/reinit.rs]: There are at least 7 places where u64 clone occurs (found them with a `id.clone()` search)
Pls remove those

File: cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs [hunk 10]
```diff
@@ -71,20 +74,22 @@ pub fn try_handle_unstake_response(
...
+                ))?,
+            )
+            .add_event(
+                Event::new("unstake_failed")
+                    .add_attribute("unstake_id", unstake_response.unstake_id.to_string())
+                    .add_attribute("lp_amount", unstake_item.lp_token_amount.to_string()),
+            );
     }
 
     UNSTAKES.save(
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs:84]: You have redundant `UNSTAKES.remove` higher on line 66 which clears the same storage slot

File: cosmwasm/contracts/yield-ward/src/execute/unstake.rs [hunk 11]
```diff
@@ -11,16 +11,6 @@ pub fn try_init_unstake(
     token_denom: TokenDenom,
     lpt_amount: Uint128,
 ) -> Result<Response, ContractError> {
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/unstake.rs:13]: Lacks the `lpt_amount != 0` check

**Review by optifat — CHANGES_REQUESTED** [review]

File: cosmwasm/contracts/yield-ward/src/execute/unstake.rs (hunk 11 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/unstake.rs:13]: Check for this is presented out of the function:
```rust
match action_msg {
    Cw20ActionMsg::Unstake => {
        if msg.amount.is_zero() {
            return Err(ContractError::ZeroTokenAmount);
        }
        let deposit_token_denom =
            find_deposit_token_denom_by_lpt_address(deps.as_ref(), &info.sender)?;
        try_init_unstake(deps, env, user, deposit_token_denom, msg.amount)
    }
}
```

File: cosmwasm/contracts/yield-ward/src/query.rs (hunk 4 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/query.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/reinit.rs (hunk 5 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/reinit.rs:88]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/common.rs (hunk 6 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/common.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/receive_cw20.rs (hunk 7 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/receive_cw20.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/response_stake.rs (hunk 8 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/response_stake.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/reinit.rs (hunk 9 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/reinit.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs (hunk 10 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs:84]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs (hunk 1 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/response_unstake.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/stake.rs (hunk 2 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/execute/stake.rs:19]: fixed

File: cosmwasm/contracts/yield-ward/src/helpers.rs (hunk 3 above)
Comment (Overseven) [reply cosmwasm/contracts/yield-ward/src/helpers.rs]: fixed

File: cosmwasm/contracts/yield-ward/src/execute/reinit.rs [hunk 12]
```diff
@@ -84,33 +78,46 @@ pub fn handle_reinit(
...
         ));
     }
+    if info.funds.first().unwrap().amount.is_zero() {
+        return Err(ContractError::CustomError(
+            "Reinit message must have non-zero token amount".to_string(),
+        ));
+    }
 
-    let (token_denom, token_config) =
+    let coin = info.funds.first().unwrap();
```
Comment (optifat) [inline cosmwasm/contracts/yield-ward/src/execute/reinit.rs]: getting coin twice

**Review by optifat — APPROVED** [review]

**Pull request merged**